pkg/trtl/replica/replica.go

```diff
@@ -0,0 +1,396 @@
+func New(conf config.Config, db *honu.DB, replicatedNamespaces []string) (*Service, error) {
+	if err := conf.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		conf:                 conf.Replica,
+		mtls:                 conf.MTLS,
+		db:                   db,
+		aestop:               make(chan struct{}),
+		replicatedNamespaces: replicatedNamespaces,
+	}, nil
+}
+
+// Shutdown the replica server (stops the anti-entropy go-routine)
+func (r *Service) Shutdown() error {
+	// If anti-entropy is enabled, send a stop signal to it. Do not send the signal if
+	// not enabled, otherwise Shutdown() will block forever and cause a deadlock.
+	if r.conf.Enabled && r.aestop != nil {
+		r.aestop <- struct{}{}
+	}
+	return nil
+}
+
+//===========================================================================
+// Gossip (server-side) Methods
+//===========================================================================
+
+// Gossip implements bilateral anti-entropy: during a Gossip session the initiating
+// replica pushes updates to the remote peer and pulls requested changes. Using
+// bidirectional streaming, the initiating peer sends data-less sync messages with
+// the versions of objects it stores locally. The remote replica then responds with
+// data if its local version is later or sends a sync message back requesting the
+// data from the initiating replica if its local version is earlier. No exchange
+// occurs if both replicas have the same version. At the end of a gossip session,
+// both replicas should have synchronized and have identical underlying data stores.
+func (r *Service) Gossip(stream replica.Replication_GossipServer) (err error) {
+	// Set up the log context for consistent logging
+	// TODO: get remote peer information from mTLS context and add to logging.
+	logctx := log.With().
+		Str("service", "gossip").
+		Bool("remote", true).
+		Logger()
+
+	// If anti-entropy is not enabled on this replica, reject the request so that the
+	// underlying database does not get modified. This method should not be served in
+	// maintenance mode, so no need to check that state.
+	if !r.conf.Enabled {
+		logctx.Debug().Msg("rejecting gossip request: anti-entropy not enabled")
+		return status.Error(codes.FailedPrecondition, "anti-entropy not enabled on remote")
+	}
+
+	// Create a stream sender to ensure that both phase1 and phase2 go routines can send
+	// messages concurrently without violating the grpc semantic that only one go
+	// routine can send messages on the stream.
+	// NOTE: newStreamSender calls wg.Add(1) and runs the sender go routine.
+	wg := new(sync.WaitGroup)
+	sender := newStreamSender(wg, logctx, stream)
+
+	// Start phase 1: receive object version vectors from the initiator. This go routine
+	// will kick off phase 2: sending unchecked versions back to the initiator.
+	wg.Add(1)
+	go r.remotePhase1(stream.Context(), wg, logctx, stream, sender)
```

Comment:
I wish we had better names for phase1 and phase2. I feel like we have a mutual understanding of what phase1 and phase2 mean after our discussions but it might be difficult for others to understand without the accompanying diagram. Unfortunately, I don't have any better suggestions right now.